### PR TITLE
[DI] add syntax to stack decorators

### LINKED ISSUE #30599: [DI] Decorate the service directly instead of override
**Description**
Add possibility to decorate the service directly instead of override.

**Example**

override:

```yaml
services:
    ham_burger:
        class: Burger

    cheese_burger:
        class: Burger

    ham.ham_decorator:
        class: HamDecorator
        decorate: ham_burger

    cheese.cheese_decorator:
        class: CheeseDecorator
        decoration_priority: 2
        decorate: cheese_burger

    ham.cheese_decorator:
        class: CheeseDecorator
        decoration_priority: 1
        decorate: cheese_burger
```

decorate :

```yaml
services:
    ham_burger:
        class: Burger
        decorators:
            - HamDecorator

    cheese_burger:
        class: Burger
        decorators:
            - CheeseDecorator
            - HamDecorator
```

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #30599
| License       | MIT
| Doc PR        | -

Declare this:
```yaml
services:
    my_stack_of_decorators:
        stack:
            - class: App\ExternalDecorator
            - class: App\InternalDecorator
            - class: App\DecoratoredClass
```

And get this:
image

The PR is now ready with support for Yaml, XML and the PHP-DSL. It needs #36388, #36392 and #36389 to pass, and relates to #36390 to be DX-friendly.

The new syntax now supports composable stacks - i.e stack you can reuse in the middle of another stack.

RIP middleware, simple decorators FTW :)

From the test cases:
```yaml
services:
    reusable_stack:
        stack:
            - class: stdClass
              properties:
                  label: A
                  inner: '@.inner'
            - class: stdClass
              properties:
                  label: B
                  inner: '@.inner'

    concrete_stack:
        stack:
            - parent: reusable_stack
            - class: stdClass
              properties:
                  label: C
```

This will create a service similar to:
```php
(object) [
    'label' => 'A',
    'inner' => (object) [
        'label' => 'B',
        'inner' => (object) [
             'label' => 'C',
        ]
    ],
];
```

When used together with autowiring, this is enough to declare a stack of decorators:
```yaml
services:
    my_processing_stack:
        stack:
            - App\ExternalDecorator: ~
            - App\InternalDecorator: ~
            - App\TheDecoratedClass: ~
```

See fixtures for the other configuration formats.

See also https://twitter.com/nicolasgrekas/status/1248198573998604288

Todo:
- [x] rebase on top of #36388, #36392 and #36389 once they are merged
- [x] test declaring deeper nested stacks

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 1]
```diff
@@ -342,6 +342,25 @@ private function parseDefinition(string $id, $service, string $file, array $defa
...
+            $stackId = null;
+
+            foreach (array_reverse($service['stack'], true) as $i => $service) {
+                if (null === $stackId) {
+                    $stackId = $id;
+                } elseif (!\is_array($service)) {
+                    throw new InvalidArgumentException(sprintf('A stacked service definition must be an array but "%s" found at position "%s" for service "%s" in "%s". Check your YAML syntax.', get_debug_type($service), $i, $id, $file));
+                } else {
+                    $stackId = '.stack.'.$id.'.'.$i;
+                    $service['decorates'] = $id;
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:355]: manually configuring `decorates` should be disallowed for stacked services, isnt it?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Ppl will figure out it doesn't make sense to define this in the middle of a stack. I'd better not add code to deal with this.

**Comment by nicolas-grekas** [comment]
The PR is now ready with support for Yaml, XML and the PHP-DSL. It needs #36388, #36392 and #36389 to pass, and relates to #36390 to be DX-friendly.

The new syntax now supports composable stacks - i.e stack you can reuse in the middle of another stack.

RIP middleware, simple decorators FTW :)

From the test cases:
```yaml
services:
    abstract_stack:
        stack:
            - class: stdClass
              abstract: true
              properties:
                  label: A
                  inner: '@.inner'
            - class: stdClass
              properties:
                  label: B
                  inner: '@.inner'

    concrete_stack:
        stack:
            - parent: abstract_stack
            - class: stdClass
              properties:
                  label: C
```

This will create a service similar to:
```php
(object) [
    'label' => 'A',
    'inner' => (object) [
        'label' => 'B',
        'inner' => (object) [
             'label' => 'C',
        ]
    ],
];
```

When used together with autowiring, this is enough to declare a stack of decorators:
```yaml
services:
    my_processing_stack:
        - App\ExternalDecorator: ~
        - App\InternalDecorator: ~
        - App\TheDecoratedClass: ~
```

See fixtures for the other configuration formats.

**Comment by dragoonis** [comment]
can you share details on why the community would switch from middleware to decorator stack pattern?

**Comment by nicolas-grekas** [comment]
@dragoonis everyone in the community will decide what they want, there is no one-size-fits-all solution. But this PR fills the gap between a middleware system and a stack of decorators. This brings "ease of stacking" to both styles. If that was the main benefit of a middleware system, this PR makes in less relevant - for ppl using the Symfony DI component of course.

**Comment by jeremyFreeAgent** [comment]
Today I do that stack part by myself. Can’t wait to use it that way.

**Comment by nicolas-grekas** [comment]
This PR is ready and I'm quite happy about it!

Here are some implementation details and a future opening:

## How does this work?

When a stack `foo` is declared, loaders create a `ChildDefinition` with the chain of decorators set as its *arguments*. The child definition is also tagged with a new `container.stack` tag. Like aliases, stacks support two attributes: `public` and `deprecated`.

This means it's possible to define a stack without using a loader, by following the above convention (all other attributes of the `ChildDefinition` would be ignored). The benefit of using a new syntax is that it makes defining the chain compatible with "defaults" set in the config file.

About the chain itself, it must be a list of either services, aliases or references. Alias and references are used undifferentiated to insert another stack definition in the middle of a new one. Both styles are needed to fit each kind of loader (yaml&xml can do only aliases and php-dsl does `ref()`).

This is already quite powerful. But there's more!

There's a second way to insert a stack inside a stack: using a "parent" in the chain. Such an insertion point will override the top-most definition in the embedded stack. This means one can also *configure* a stack while including it! All attributes of child definitions work here, it all depends on what the stack is about and only you will know about this.

## What could come next?

I've been talking with @dunglas and @pamil on Twitter and Slack and they asked me about inserting or removing frames inside an existing stack.

For now, this would be reasonably easy using a compiler pass.

Using configuration, we could also imagine new tags. E.g. one to allow filtering a frame while embedding a stack, another to allow replacing a frame by a new one, another for inserting a frame before/after another one. I didn't mention it, but frames can be named, this could help.

All this could replace an event dispatcher in some situations I believe. But that's hypotheses for the future.

Please give this a try for now :)

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**